### ISSUE TITLE
Please re-allow custom attributes in colDef; suppress console messages.

### ISSUE THREAD

**Issue by jjurczewsky** [ag-grid/ag-grid#2320]
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report => search github for a similar issue or PR before submitting
[x ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/ceolter/ag-grid/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
If a colDef property is specified that agGrid is not aware of, a console.warn message is automatically generated.  The message reads "ag-grid: invalid colDef property 'unknown_prop_name' did you mean any of these: 

**Expected behavior**
An unknown colDef property would be silently ignored.

**Minimal reproduction of the problem with instructions**
Add a colDef property     doNotExport: true     to any column definition.

**What is the motivation / use case for changing the behavior?**
Because the colDef is sent in nearly all callbacks, we use custom attributes in the colDef.  For example, a colDef of    doNotExport: true   tells our code not to dump this column to Excel.  In past versions, this worked great: ag-grid silently ignored anything it didn't 'understand'.  Now we get console.warn messages that, to my knowledge, cannot be suppressed.  This doesn't look good in our production code.

As a further note, it is my opinion that all ag-grid console.xxxx messages should be suppressable, ideally through a single setting.  I've noticed many messages that get generated (e.g., ag-Grid.Context: bean [Object O created) that simply clutter the console.

Bottom line: a) allow custom properties to be set colDef, and b) please allow a way to suppress all ag-grid console.xxxx messages.

**Please tell us about your environment:**
Windows 10, phpStorm, yarn, apache

* **ag-Grid version:** 17.0.0

* **Browser:**
Al

* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
all

**Comment by flybayer** [comment]
I add my request to this issue also. 

I'm using `ag-grid-vue` which adds an `__ob__` property to the column definition. This causes ag-grid to spew out a warning for every single column.

**Comment by voyti** [comment]
+1 for this  - is there any technical reason not to keep custom params in colDefs?

**Comment by espenkm** [comment]
For me I got rid of the messages by setting
GridOptions {
            suppressPropertyNamesCheck: true
}

which may or may not be a good idea

**Comment by zweeen** [comment]
Thanks @espenkm your solution worked for me :)

**Comment by sagarloona4180** [comment]
yes, same things also work for me.

**Comment by BrianHVB** [comment]
+1

Same use case as jjurczewsky. We are using React and exporting grid data to a custom database format. In keeping with React's declarative approach, we have a "suppressExport" attribute that are added to custom column components such as buttons, checkboxes, etc.

Suppressing the warning works, but is a slippery slope. It seems that a better default would be to have a `verbose` settings for those circumstances where a developer mistypes or misspells a column property and can't figure out why it isn't having any effect.

**Comment by kavika-1** [comment]
How about a single named property throughout like 'userData'?  Then we are responsible for anything under that, and warnings can still pop.  That might create a complex mess for libraries that enhance the grid, but for app writers it would be nice.

**Comment by makinggoodsoftware** [comment]
Hi,

It seems like you guys found the property to ignore the property check, I am going to close this issue.

Thanks

**Comment by ondrakoupil** [comment]
Actually I find @kavika-1 's idea to be a very good one. The property names check can be helpful, while the possibility to add some custom data to column definition is also useful in many cases. By adding one new property to ColDef that can hold any arbitrary data, both features could be used and benefited from. And also, Typescript wouldn't complain (the new property would be simply typed as "any"). 
Maybe this could be added in some future version?

**Comment by michaeljohnston-chq** [comment]
@makinggoodsoftware Suppressing the messages still causes custom properties to be removed from the definition, though - why close the issue when the main request is not addressed?

**Comment by asifsha** [comment]
Can you please reopen this issue? as the problem is not solved yet, also setting the suppressPropertyNamesCheck to true is not correct in this situation, as it also hides other potential warnings
```
this.gridOptions = {
     ...
    suppressPropertyNamesCheck: true           
};

```
![image](https://user-images.githubusercontent.com/33965589/63589318-3b41b500-c5c2-11e9-83dd-9ecbae1aa97a.png)

**Comment by howlettga** [comment]
custom data property on columns would be nice. Is there any alternative right now? suppressing console warnings does nothing to TypeScript checking in my editor. Would rather not create a separate map of column Ids to my custom attributes

**Comment by NoahStahl** [comment]
Agreed, an extension property would be useful to track custom column metadata.

**Comment by KalalauEnterprises** [comment]
Please allow custom fields in colDef without warnings!!!

**Comment by jackdh** [comment]
still requesting please.

**Comment by KalalauEnterprises** [comment]
Yes,Please!!

**Comment by dvins** [comment]
@jjurczewsky Please consider reopening this issue. There are valid reasons for adding custom attributes to column definitions and simply disabling all checks via `suppressPropertyNamesCheck: true` is a horrible solution. We truly need a way to at least identify to ag-grid what our custom attributes are so that it will skip over them, but still provide validation for native properties.

**Comment by AnaSofiaGrilo96** [comment]
I also would appreciate to have a way to add custom properties to a column! Please consider it!

**Comment by thisentirecitymustbepurged** [comment]
Using `suppressPropertyNamesCheck` to lose all property checking for one or two custom column properties is simply absurd.

Another solution I'd like to suggest is for `suppressPropertyNamesCheck` to accept a list of custom property names, so 

`suppressPropertyNamesCheck: ["Alabama", "420"]`

means that `Alabama` and `420` should be legal to use.

**Comment by mwsiddeeq15** [comment]
> Using `suppressPropertyNamesCheck` to lose all property checking for one or two custom column properties is simply absurd.
> 
> Another solution I'd like to suggest is for `suppressPropertyNamesCheck` to accept a list of custom property names to be ignored by the property name checking, so
> 
> `suppressPropertyNamesCheck: ["Alabama", "420"]`
> 
> means that `Alabama` and `420` should be legal.

➕

**Comment by muenchto** [comment]
@makinggoodsoftware please reopen. The main request of this issue was not addressed and the need for a solution is still there. Thanks!

**Comment by BrentMShanahan** [comment]
@makinggoodsoftware Agreed that this needs to be reopened. Being able to store metadata with the column definition should be allowed. Whether it's a single parameter like @kavika-1 suggested or a way to suppress warnings on a specific basis like @thisentirecitymustbepurged.

**Comment by vagoel** [comment]
Please reopen this ticket.
Setting this -
```
this.gridOptions = {
     ...
    suppressPropertyNamesCheck: true           
};

```

is not a good solution.It might suppress useful warning messages.

As suggested by @kavika-1  , 


> How about a single named property throughout like 'userData'? Then we are responsible for anything under that, and warnings can still pop. That might create a complex mess for libraries that enhance the grid, but for app devs it would be nice

looks promising.

**Comment by rcketscientist** [comment]
Here's a way to have your cake and eat it to:
```
PropertyKeys.ALL_PROPERTIES = [
  ...PropertyKeys.ALL_PROPERTIES,
  'canAutoHideColumns',
  'contextMenu'
];
```
```
// Patch our custom properties into the valid properties
ColDefUtil.ALL_PROPERTIES = [
    ...ColDefUtil.ALL_PROPERTIES,
    'canAutoHide'
];
```

**Comment by cstephe** [comment]
The warnings are over and under engineered at the same time.  The intent is to help devs by finding potential mistakes in configs.  However, it makes things unnecessarily rigid and brittle.  If we add a field whose purpose is for loading up custom fields then that shows intent.  I use a helper to generate my gridOptions and if I want to extend my functionality with a custom field I only need to add customProps: [...customerProps, 'displayWhenEmpty'] to a single place.  If you wanted to do more you could check them against the known props for collisions and give a warning then.

**Comment by cstephe** [comment]
I changed my mind.  I can use something like cellRendererParams and get that from the col def.  I am using them in value formatters and this would work perfectly.

**Comment by JayantSonawane** [comment]
Hi, getting this on my console
![image](https://user-images.githubusercontent.com/55382974/130599491-41972e91-d1fd-445a-a4fe-25a191767d36.png)

can anyone help with that

**Comment by mwsiddeeq15** [comment]
> Here's a way to have your cake and eat it to:
> 
> ```
> PropertyKeys.ALL_PROPERTIES = [
>   ...PropertyKeys.ALL_PROPERTIES,
>   'canAutoHideColumns',
>   'contextMenu'
> ];
> ```
> 
> ```
> // Patch our custom properties into the valid properties
> ColDefUtil.ALL_PROPERTIES = [
>     ...ColDefUtil.ALL_PROPERTIES,
>     'canAutoHide'
> ];
> ```

Where do you import these Objects from? and where do you set them? and does this actually work to add custom columnDef properties?

**Comment by GOVYANSONG** [comment]
@mwsiddeeq15 it is located in the code file: https://github.com/ag-grid/ag-grid/blob/10f77815ae533df0dc205123bc956f9cc1e74f09/community-modules/core/src/ts/components/colDefUtil.ts

**Comment by danielelkington** [comment]
It seems that ColDefUtil is no longer exported, and thus the suggested solution no longer works.

**Comment by StephenCooper** [comment]
Thanks @muenchto for the details and your work around. I think this is probably the best approach to go with for now. 

> It seems that ColDefUtil is no longer exported, and thus the suggested solution no longer works.
This is true for v29 as we made changes in our internal use of this class that meant it no longer needed to be shared across modules. I will add it back to the exports for v29.1 to allow this work around to keep working for you as it does not seem that we will be making a change to add this feature anytime soon.

**Comment by yashwanth2714** [comment]
Hello @StephenCooper , This problem has resurfaced in the latest version 31. Can you help me resolve it?

**Comment by demc** [comment]
> Hello @StephenCooper , This problem has resurfaced in the latest version 31. Can you help me resolve it?

This is what I did:

```
import { PropertyKeys } from 'ag-grid-community';
PropertyKeys.ALL_PROPERTIES.push('foo');
```

**Comment by NiclasFalkman** [comment]
> > Hello @StephenCooper , This problem has resurfaced in the latest version 31. Can you help me resolve it?
> 
> This is what I did:
> 
> ```
> import { PropertyKeys } from 'ag-grid-community';
> PropertyKeys.ALL_PROPERTIES.push('foo');
> ```

It still complaints when I try this. Does anyone have another solution to this?

**Comment by demc** [comment]
> > > Hello @StephenCooper , This problem has resurfaced in the latest version 31. Can you help me resolve it?
> > 
> > This is what I did:
> > 
> > ```
> > import { PropertyKeys } from 'ag-grid-community';
> > PropertyKeys.ALL_PROPERTIES.push('foo');
> > ```
> 
> It still complaints when I try this. Does anyone have another solution to this?

Try this prop on the grid:

gridOptions.suppressPropertyNamesCheck

**Comment by emil-eklund** [comment]
@demc To quote an earlier message:

> Using `suppressPropertyNamesCheck` to lose all property checking for one or two custom column properties is simply absurd.
> 
> Another solution I'd like to suggest is for `suppressPropertyNamesCheck` to accept a list of custom property names to be ignored by the property name checking, so
> 
> `suppressPropertyNamesCheck: ["Alabama", "420"]`
> 
> means that `Alabama` and `420` should be legal.

**Comment by jcmnavia** [comment]
> Hello,
> 
> Thank you for bringing this up with us.
> 
> AG Grid offers a context property inside grid options to allow you to set meta-data on the grid level. https://www.ag-grid.com/javascript-data-grid/grid-options/#reference-miscellaneous-context
> 
> We can support your requirement by applying the same pattern to column definitions and adding a similar context property at the column definition level. This way you'll be able to add your custom properties there without having to disable property name checks across the grid.
> 
> We have added a feature request in our backlog to track this requirement and implement it in the future.
> 
> You can track it with this reference: **AG-11710 [API] Add a context property to column definitions to support custom meta-data on the column-level**
> 
> You can follow the status of feature requests, bugs and releases using our pipeline: https://www.ag-grid.com/ag-grid-pipeline/
> 
> **Now that this requirement is recorded in our backlog, we will soon close the GitHub issue because we will not update it when this feature is added in a future version of AG Grid. The best way to get an update on this item is to sign up for AG Grid new release notifications. Once a new AG Grid version is released, look for the item reference above (AG-XXXX) on the changelog page to see if it's resolved in that version: https://www.ag-grid.com/ag-grid-changelog/**
> 
> Thanks again for bringing this up with us - we appreciate your time!

It would be great if you update the Column documentation with this with a few quick examples, would have saved me a lot of time
I had to search first in the pipeline, then in the changelog and then directly in the PRs https://github.com/ag-grid/ag-grid/pull/8101 to know how this works

Still, great to know this exists already!!! Thanks a lot 💯

**Comment by sovetski** [comment]
From my understanding, we have 2 options

- suppressPropertyNamesCheck: for me it is not optimal as explained by @js1m 
- context: the docs is not complete for this, I don't understand anything about it and from what I see it is not on "colDef" level but ag grid options

How we are supposed to add any custom data in column level??

**Comment by kiril-matev** [comment]
Hi @sovetski,

The grid allows you to set context on both grid and individual column definition levels.

You can provide column-specific context by setting the column definition context property as documented here:
https://www.ag-grid.com/javascript-data-grid/column-properties/#reference-columns-context

We'll make a note to update the context documentation page to include both the grid and column-level context properties.

**Comment by npnsap** [comment]
Hi @kiril-matev,
  
We are using AG Grid with a couple of custom features. These features are controlled through global or page-level configurations and passed to `gridOptions` using custom properties. For us, these custom properties in `gridOptions` are not just meta-data—they control the functionality of grid. We already use `context` for passing actual meta-data related to the grid.

So, it would be great if you could keep both options and continue allowing the use of `suppressPropertyNamesCheck` instead of deprecating it and forcing the use of context. That keeps ag-grid flexible and extendable.

**Comment by agentscript** [comment]
In your AGGrid component you can add context and add any property and it should work fine. You can create extended class of ColDef or suppress warning but you will see a message in the console easiest thing to do is add the field name auditFieldName in the example below


 <AgGridReact
          ref={gridRef}
          onGridReady={onGridReady}                   
          context={{
            auditFieldName: undefined,
            enableHistory: false,          
          }}/>
